hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -566,35 +566,65 @@ def blocks(self):
     def all_block_bounds(self):
         return self.shrink_target.all_block_bounds()
 
-    def each_pair_of_blocks(self, accept_first, accept_second):
-        """Yield each pair of blocks ``(a, b)``, such that ``a.index <
-        b.index``, but only if ``accept_first(a)`` and ``accept_second(b)`` are
-        both true."""
+    def each_pair(self, base_collection, dependent_collection):
+        """Iterate over the (filtered) product of a pair of dynamically
+        changing collections.
 
-        # Iteration order here is significant: Rather than fixing i and looping
-        # over each j, then doing the same, etc. we iterate over the gap between
-        # i and j and then over i. The reason for this is that it ensures that
-        # we try a different value for i and j on each iteration of the inner
-        # loop. This stops us from stalling if we happen to hit on a value of i
-        # where nothing useful can be done.
-        #
-        # In the event that nothing works, this doesn't help and we still make
-        # the same number of calls, but by ensuring that we make progress we
-        # have more opportunities to make shrinks that speed up the tests or
-        # that reduce the number of viable shrinks at the next gap size because
-        # we've lowered some values.
-        offset = 1
-        while offset < len(self.blocks):
+        Logically equivalent to:
+
+        .. code-block: python
+
+            outer = base_collection()
+            for a in outer:
+                for b in dependent_collection(a):
+                    yield (a, b)
+
+        except that:
+
+            * If the collection changes due to shrinking, the
+              iteration will be "fixed" to use the updated collection.
+              The result will be that every pair that occurs after a change
+              has happened will be one that could have arisen if the iteration
+              started with the current value.
+            * The iteration order is different, in a way that is designed
+              to reduce the likelihood of the shrinker "stalling", by ensuring
+              that both a and b change on each loop iteration.
+        """
+
+        def recalc():
+            base = base_collection()
+            return base, [None] * len(base)
+
+        outer, children = recalc()
+
+        offset = 0
+        while True:
             i = 0
-            while i + offset < len(self.blocks):
-                j = i + offset
-                block_i = self.blocks[i]
-                block_j = self.blocks[j]
-                if accept_first(block_i) and accept_second(block_j):
-                    yield (block_i, block_j)
+            any_in_offset = False
+            while i < len(outer):
+                if children[i] is None:
+                    children[i] = dependent_collection(outer[i])
+                inner = children[i]
```

Comment:
Since we're using `None` as a sentinel, an `assert inner is not None` will guard against it accidentally being used as a legitimate value somehow.

Reply:
I thought about it but it will immediately fail when we call `len(inner)` so this doesn't actually add any debugging benefit.